src/pull.rs

```diff
@@ -1,42 +1,173 @@
 use std::f32;
 
 use nalgebra;
-use nalgebra::{Point2, Point3, Rotation3, Vector3};
+use nalgebra::{Point3, Rotation3, Vector3};
 
 use crate::convert::cast_usize;
-use crate::geometry::{self, Face, Geometry, Plane, UnorientedEdge};
-
-/// Compute barycentric coordinates of point P in triangle A, B, C. Returns None
-/// for degenerate triangles.
-pub fn compute_barycentric_coords(
-    a: Point2<f32>,
-    b: Point2<f32>,
-    c: Point2<f32>,
-    p: Point2<f32>,
-) -> Option<Point3<f32>> {
-    let ab = b - a;
-    let ac = c - a;
-    let pa = a - p;
-    let xs = Vector3::new(ac.x, ab.x, pa.x);
-    let ys = Vector3::new(ac.y, ab.y, pa.y);
-    let ortho = xs.cross(&ys);
-    if f32::abs(ortho.z) < 1.0 {
-        None
+use crate::geometry;
+use crate::mesh::{Face, Mesh, UnorientedEdge};
+use crate::plane::Plane;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct PulledPointWithDistance {
+    pub point: Point3<f32>,
+    pub distance: f32,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct PointOnLine {
+    pub clamped: Point3<f32>,
+    pub unclamped: Point3<f32>,
+}
+
+/// Pulls arbitrary point to an arbitrary line.
+///
+/// Returns both, a clamped point on the line and an unclamped points on an
+/// endless ray.
+///
+/// https://stackoverflow.com/questions/3120357/get-closest-point-to-a-line
+pub fn pull_point_to_line(
+    point: &Point3<f32>,
+    line_start: &Point3<f32>,
+    line_end: &Point3<f32>,
+) -> PointOnLine {
+    if is_point_on_line_clamped(point, line_start, line_end) {
+        return PointOnLine {
+            clamped: *point,
+            unclamped: *point,
+        };
+    }
+    let start_to_point = point - line_start;
+    let start_to_end = line_end - line_start;
+
+    let square_magnitude_start_to_end = nalgebra::distance_squared(line_start, line_end);
+    let start_to_point_dot_start_to_end = start_to_point.dot(&start_to_end);
+    let line_parameter = start_to_point_dot_start_to_end / square_magnitude_start_to_end;
+
+    let closest_point_on_line_unclamped =
+        Point3::from(line_start.coords.lerp(&line_end.coords, line_parameter));
+    let closest_point_on_line_clamped = if line_parameter < 0.0 {
+        *line_start
+    } else if line_parameter > 1.0 {
+        *line_end
     } else {
-        Some(Point3::new(
-            1.0 - (ortho.x + ortho.y) / ortho.z,
-            ortho.y / ortho.z,
-            ortho.x / ortho.z,
-        ))
+        closest_point_on_line_unclamped
```

Comment:
right

Reply:
No, this is just a very confusing diff. The deleted section you see is from another function.